File: app/src/main/java/com/spproject/audiowatermark/WavUtils.kt
package com.spproject.audiowatermark

import android.content.Context
import java.io.File
import java.io.FileOutputStream

/**
 * WAV file I/O — reads and writes 44100 Hz / mono / 16-bit PCM.
 *
 * Required host-song format (convert with ffmpeg if needed):
 *   ffmpeg -i your_song.mp3 -ar 44100 -ac 1 -sample_fmt s16 host_song.wav
 * Then place host_song.wav in app/src/main/res/raw/.
 */
object WavUtils {

    // ── Load ─────────────────────────────────────────────────────────────

    /**
     * Loads a WAV resource from res/raw into a normalized [-1.0, 1.0] DoubleArray.
     *
     * Supports mono and stereo input; stereo is downmixed to mono by averaging
     * the two channels.  Sample rate MUST be [Config.SAMPLE_RATE] (validated).
     *
     * @param context Android context used to open the resource.
     * @param resId   R.raw.host_song (or any other WAV resource ID).
     * @throws IllegalArgumentException if the file is not a valid WAV, not 16-bit
     *                                  PCM, or has the wrong sample rate.
     */
    fun loadWavAsDoubles(context: Context, resId: Int): DoubleArray {
        val bytes = context.resources.openRawResource(resId).use { it.readBytes() }

        // Parse RIFF/WAVE header iterating over chunks
        require(bytes.size >= 12 &&
                String(bytes, 0, 4, Charsets.US_ASCII) == "RIFF" &&
                String(bytes, 8, 4, Charsets.US_ASCII) == "WAVE") {
            "Not a RIFF/WAVE file"
        }

        var pos = 12
        var numChannels = 1
        var sampleRate = 0
        var bitsPerSample = 16
        var dataOffset = -1
        var dataSize = -1

        while (pos + 8 <= bytes.size) {
            val chunkId = String(bytes, pos, 4, Charsets.US_ASCII)
            val chunkSize = readLE32(bytes, pos + 4)
            when (chunkId) {
                "fmt " -> {
                    val audioFormat = readLE16(bytes, pos + 8)   // 1 = PCM
                    require(audioFormat == 1) {
                        "Only PCM WAV is supported (audioFormat=$audioFormat). " +
                        "Reconvert with ffmpeg: ffmpeg -i input.mp3 -ar 44100 -ac 1 -sample_fmt s16 host_song.wav"
                    }
                    numChannels = readLE16(bytes, pos + 10)
                    sampleRate  = readLE32(bytes, pos + 12)
                    bitsPerSample = readLE16(bytes, pos + 22)
                }
                "data" -> {
                    dataOffset = pos + 8
                    dataSize   = chunkSize
                }
            }
            // Chunks are padded to even byte boundaries
            pos += 8 + chunkSize + (chunkSize and 1)
            if (dataOffset != -1) break
        }

        require(dataOffset != -1)  { "No 'data' chunk found — is this a valid WAV?" }
        require(bitsPerSample == 16) {
            "Expected 16-bit PCM, got ${bitsPerSample}-bit. Reconvert with ffmpeg -sample_fmt s16."
        }
        require(sampleRate == Config.SAMPLE_RATE) {
            "Host WAV is $sampleRate Hz but Config.SAMPLE_RATE is ${Config.SAMPLE_RATE} Hz. " +
            "Reconvert: ffmpeg -i song.mp3 -ar ${Config.SAMPLE_RATE} -ac 1 -sample_fmt s16 host_song.wav"
        }

        val totalFrames = dataSize / (numChannels * 2)   // 2 bytes per 16-bit sample
        val doubles = DoubleArray(totalFrames)
        var bi = dataOffset
        for (i in 0 until totalFrames) {
            if (numChannels == 1) {
                doubles[i] = readLE16Signed(bytes, bi) / 32768.0
                bi += 2
            } else {
                // Downmix stereo → mono by averaging L and R channels
                val left  = readLE16Signed(bytes, bi)
                val right = readLE16Signed(bytes, bi + 2)
                doubles[i] = (left + right) / 2.0 / 32768.0
                bi += numChannels * 2
            }
        }
        return doubles
    }

    // ── Write ─────────────────────────────────────────────────────────────

    /**
     * Writes normalized [-1.0, 1.0] [samples] to a 44100 Hz / mono / 16-bit
     * PCM WAV file at [outFile].
     *
     * Useful for saving the watermarked audio to internal storage so you can
     * inspect it offline (e.g. open in Audacity via adb pull).
     *
     * @param samples  Normalized PCM — e.g. the output of [Embedder.embed].
     * @param outFile  Destination file; will be created/overwritten.
     */
    fun writeWav(samples: DoubleArray, outFile: File) {
        val numSamples = samples.size
        val byteRate   = Config.SAMPLE_RATE * 2         // 1 ch × 2 bytes/sample
        val dataSize   = numSamples * 2                 // bytes in data chunk
        val fileSize   = 36 + dataSize                  // total RIFF size - 8

        FileOutputStream(outFile).use { fos ->
            // RIFF header
            fos.write("RIFF".toByteArray(Charsets.US_ASCII))
            fos.writeLE32(fileSize)
            fos.write("WAVE".toByteArray(Charsets.US_ASCII))

            // fmt  chunk (16 bytes)
            fos.write("fmt ".toByteArray(Charsets.US_ASCII))
            fos.writeLE32(16)                            // chunk size
            fos.writeLE16(1)                             // PCM
            fos.writeLE16(1)                             // mono
            fos.writeLE32(Config.SAMPLE_RATE)
            fos.writeLE32(byteRate)
            fos.writeLE16(2)                             // block align
            fos.writeLE16(16)                            // bits per sample

            // data chunk
            fos.write("data".toByteArray(Charsets.US_ASCII))
            fos.writeLE32(dataSize)
            val buf = ByteArray(2)
            for (s in samples) {
                val v = (s * 32767.0).toInt().coerceIn(-32768, 32767)
                buf[0] = (v and 0xFF).toByte()
                buf[1] = ((v shr 8) and 0xFF).toByte()
                fos.write(buf)
            }
        }
    }

    // ── LE read helpers ──────────────────────────────────────────────────

    private fun readLE16(b: ByteArray, offset: Int): Int =
        (b[offset].toInt() and 0xFF) or ((b[offset + 1].toInt() and 0xFF) shl 8)

    /** Reads a signed 16-bit little-endian integer. */
    private fun readLE16Signed(b: ByteArray, offset: Int): Int {
        val lo = b[offset].toInt() and 0xFF
        val hi = b[offset + 1].toInt()          // sign-extending toInt()
        return (hi shl 8) or lo
    }

    private fun readLE32(b: ByteArray, offset: Int): Int =
        (b[offset].toInt() and 0xFF) or
        ((b[offset + 1].toInt() and 0xFF) shl 8) or
        ((b[offset + 2].toInt() and 0xFF) shl 16) or
        ((b[offset + 3].toInt() and 0xFF) shl 24)

    // ── LE write helpers ─────────────────────────────────────────────────

    private fun FileOutputStream.writeLE16(v: Int) {
        write(v and 0xFF)
        write((v shr 8) and 0xFF)
    }

    private fun FileOutputStream.writeLE32(v: Int) {
        write(v and 0xFF)
        write((v shr 8) and 0xFF)
        write((v shr 16) and 0xFF)
        write((v shr 24) and 0xFF)
    }
}

File: working codes/lib/dsp/wav_utils.dart
import 'dart:io';
import 'dart:typed_data';

class WavAudio {
  final int sampleRate;
  final int channels;
  final List<double> samples;

  WavAudio({
    required this.sampleRate,
    required this.channels,
    required this.samples,
  });
}

WavAudio readWav(String path) {
  final bytes = File(path).readAsBytesSync();
  return readWavBytes(bytes);
}

WavAudio readWavBytes(Uint8List bytes) {
  final data = ByteData.sublistView(bytes);

  if (bytes.length < 12 || String.fromCharCodes(bytes.sublist(0, 4)) != 'RIFF' ||
      String.fromCharCodes(bytes.sublist(8, 12)) != 'WAVE') {
    throw ArgumentError('Not a valid WAV byte buffer');
  }

  int? fmtOffset;
  int? dataOffset;
  int? dataSize;

  int offset = 12;
  while (offset + 8 <= bytes.length) {
    final id = String.fromCharCodes(bytes.sublist(offset, offset + 4));
    final chunkSize = data.getUint32(offset + 4, Endian.little);
    if (id == 'fmt ') {
      fmtOffset = offset + 8;
    } else if (id == 'data') {
      dataOffset = offset + 8;
      dataSize = chunkSize;
    }
    offset += 8 + chunkSize + (chunkSize % 2);
  }

  if (fmtOffset == null || dataOffset == null || dataSize == null) {
    throw ArgumentError('WAV byte buffer is missing required fmt/data chunks');
  }

  final audioFormat = data.getUint16(fmtOffset, Endian.little);
  if (audioFormat != 1) {
    throw ArgumentError('Only PCM WAV files are supported');
  }

  final channels = data.getUint16(fmtOffset + 2, Endian.little);
  final sampleRate = data.getUint32(fmtOffset + 4, Endian.little);
  final bitsPerSample = data.getUint16(fmtOffset + 14, Endian.little);

  if (bitsPerSample != 16) {
    throw ArgumentError('Only 16-bit PCM WAV files are supported');
  }

  final sampleCount = dataSize ~/ (bitsPerSample ~/ 8);
  final monoSamples = <double>[];

  for (int index = 0; index < sampleCount; index += channels) {
    var sum = 0.0;
    for (int ch = 0; ch < channels; ch++) {
      final sampleIndex = dataOffset + ((index + ch) * 2);
      final pcm = data.getInt16(sampleIndex, Endian.little);
      sum += pcm / 32768.0;
    }
    monoSamples.add(sum / channels);
  }

  return WavAudio(sampleRate: sampleRate, channels: 1, samples: monoSamples);
}

void writeWav(String path, WavAudio audio) {
  final dataSize = audio.samples.length * 2;
  final buffer = BytesBuilder();
  const chunkSize = 16;

  buffer.add('RIFF'.codeUnits);
  buffer.add((40 + dataSize).toBytesLE(4));
  buffer.add('WAVE'.codeUnits);
  buffer.add('fmt '.codeUnits);
  buffer.add(chunkSize.toBytesLE(4));
  buffer.add((1).toBytesLE(2));
  buffer.add(audio.channels.toBytesLE(2));
  buffer.add(audio.sampleRate.toBytesLE(4));
  buffer.add((audio.sampleRate * audio.channels * 2).toBytesLE(4));
  buffer.add((audio.channels * 2).toBytesLE(2));
  buffer.add((16).toBytesLE(2));
  buffer.add('data'.codeUnits);
  buffer.add(dataSize.toBytesLE(4));

  for (final sample in audio.samples) {
    final clamped = sample.clamp(-1.0, 1.0);
    final value = (clamped * 32767.0).round().clamp(-32768, 32767);
    buffer.add(value.toBytesLE(2));
  }

  File(path).writeAsBytesSync(buffer.toBytes());
}

extension _IntBytes on int {
  List<int> toBytesLE(int byteLength) {
    final bytes = Uint8List(byteLength);
    final b = ByteData.sublistView(bytes);
    switch (byteLength) {
      case 2:
        b.setInt16(0, this, Endian.little);
        break;
      case 4:
        b.setInt32(0, this, Endian.little);
        break;
      default:
        throw ArgumentError('Unsupported byte length $byteLength');
    }
    return bytes;
  }
}
